docs/DeveloperGuide.md

```diff
@@ -1,29 +1,204 @@
 # Developer Guide
```

Comment:
It would be good to include a table of contents! I found it hard to navigate through the document and to figure out what particular section I'm looking at.